Martin's review (#158)
NTAP/rfc8312bis: draft-ietf-tcpm-rfc8312bis.md

    target: "https://raw.githubusercontent.com/bbriscoe/cubic-reno/main/creno_tr.pdf"
    seriesinfo: RFC Editor, please replace this URL with the permanent arXiv one

>   Average-Cubic-Window:

Reviewer: Couldn't we just add this proof to the I-D in an appendix instead, rather than linking to a PDF that might go away?

Author: that is a good idea. I don't know how to write all the math equations, can you help with adding the proof?

Reviewer: I'll try; will prepare a separate PR.

Author: Do you want me to remove the reference now or would you remove it later?

Reviewer: Let's wait until later. I asked @lisongxu for the Latex sources for the PDF, which will make my job easier.

Reviewer: See #160. Should I merge that now, so you can refer to it in this PR?

Reviewer 2: sorry for my late response. I have emailed the latex source code to both of you.

Author: thank Lisong and Lars. I have now changed the reference to section added in #160.

when they operate outside the Reno-friendly region.

This notion of a linear throughput ratio is similar to that of Reno
under an asynchronous loss model where flows with different RTTs
have the same packet loss rate but experience loss events at different
times. However, under a synchronous loss model, where flows with different
RTTs experience loss events at the same time but have different packet loss
rates, the throughput ratio of Reno flows with different RTTs is quadratically
proportional to the inverse of their RTT ratio {{XHR04}}.

CUBIC always ensures a linear throughput ratio independent of the
loss environment. This is an improvement over Reno.

> While there is no consensus on the optimal throughput ratios for

Reviewer: ```suggestion
While there is no consensus on the optimal throughput ratio for
```

different RTT flows, over wired Internet paths, use of
a linear throughput ratio seems more reasonable than equal throughputs
(i.e., the same throughput for flows with different RTTs) or a
higher-order throughput ratio (e.g., a quadratical throughput ratio of

> Reno under synchronous loss environments).

Reviewer: ```suggestion
Reno in synchronous loss environments).
```

when they operate outside the Reno-friendly region.

This notion of a linear throughput ratio is similar to that of Reno

> under an asynchronous loss model where flows with different RTTs

Reviewer: ```suggestion
under an asynchronous loss model, where flows with different RTTs
```

3. The convex region, if CUBIC is not in the Reno-friendly region and
   *cwnd* is greater than *W<sub>max</sub>*.

To summarize, CUBIC computes both W<sub>cubic</sub>(*t*) and

> *W<sub>est</sub>* {{Reno-friendly}} on receiving a new ACK

Reviewer: What is `{{Reno-friendly}}` supposed to render as?

Author: It would render as Section 4.3

Reviewer: Shouldn't you do something like this?
```suggestion
*W<sub>est</sub>* (see {{Reno-friendly}}) on receiving a new ACK
```

3. The convex region, if CUBIC is not in the Reno-friendly region and
   *cwnd* is greater than *W<sub>max</sub>*.

To summarize, CUBIC computes both W<sub>cubic</sub>(*t*) and
*W<sub>est</sub>* (see {{Reno-friendly}}) on receiving a new ACK

> in congestion avoidance and choose the higher of the two values.

Reviewer: ```suggestion
in congestion avoidance and chooses the larger of the two values.
```

*W<sub>est</sub>* is set equal to *cwnd<sub>epoch</sub>* at the start
of the congestion avoidance stage. After that, on every new ACK,
*W<sub>est</sub>* is updated using {{eq4}}. Note that this equation

> is for a connection where Appropriate Byte Counting (ABC)

Reviewer 3: There's no need to introduce the named concept of ABC here. Sec 3.1 5681 RECOMMENDS that you count the bytes until it reaches cwnd and then increment by one MSS, but you MAY do cwnd += SMSS*SMSS/cwnd.

I would recommend that you essentially recreate this language: RECOMMEND that you count bytes, but you MAY use Eq. 4.

Author: What RFC 5681 recommends with `cwnd += SMSS*SMSS/cwnd` is different from Equation 4. in RFC8312bis as the latter uses segments_acked/cwnd where segments_acked is number of MSS sized segments. RFC 5681 doesn't consider no of segments/bytes acked but just increments by `SMSS*SMSS/cwnd` which is not the same as Eq 4.

Counting segments vs bytes are one and the same thing if segments are counted as 1 full MSS which is how segments_acked is defined in the RFC.

The other thing about counting bytes until it reaches cwnd is not necessarily better than increasing cwnd on every ACK. Both the methods should yield the same result after an RTT if segments_acked is counted correctly.

Reviewer 3: Yes, I understand that they're different.

The high-level comment is to actually write down what you want senders to do, rather than rely on confusing references to 5681 or 3465, especially, when those references aren't precisely describing your intent.

If there's a particular behavior that you can clearly and unambiguously reference to save yourself some text, that's fine. Otherwise, just be specific.

Author: ok, I have tried to clean up the text by removing the confusing reference. Can you review?

*W<sub>est</sub>* is set equal to *cwnd<sub>epoch</sub>* at the start
of the congestion avoidance stage. After that, on every new ACK,
*W<sub>est</sub>* is updated using {{eq4}}. Note that this equation
uses *segments_acked* and *cwnd* is measured in segments. An implementation

> that measures *cwnd* in bytes should adjust the equation accordingly using

Reviewer: ```suggestion
that measures *cwnd* in bytes SHOULD adjust the equation accordingly using
```
?

Author: It is not really SHOULD in IETF terminology, I think. Its just an advice for implementations to convert from segments to bytes.

Reviewer: OK, then let's leave it.

it can achieve the same congestion window increment rate as Reno,
which uses AIMD(1, 0.5).

In the next two sections, CUBIC is not in the Reno-friendly region and

> uses the window increase function described in {{win-inc}}. Although the

Reviewer: ```suggestion
uses the window increase function described in {{win-inc}}. Although 
```

it can achieve the same congestion window increment rate as Reno,
which uses AIMD(1, 0.5).

> In the next two sections, CUBIC is not in the Reno-friendly region and

Reviewer: ```suggestion
The next two sections assume that CUBIC is not in the Reno-friendly region and
```

With a deterministic loss model where the number of packets between
two successive packet losses is always *1/p*, CUBIC always operates
with the concave window profile, which greatly simplifies the

> performance analysis of CUBIC. The average window size of CUBIC ({{proof-avg-window}})

Reviewer: ```suggestion
performance analysis of CUBIC. The average window size of CUBIC (see {{proof-avg-window}})
```
